components/script/stylesheet_loader.rs

```diff
@@ -119,6 +124,118 @@ impl StylesheetContext {
+        }
+        let link = element.downcast::<HTMLLinkElement>().unwrap();
+        !self
+            .request_generation_id
+            .is_none_or(|generation| generation == link.get_request_generation_id())
+    }
+
+    fn decrement_load_and_render_blockers(&self, owner: &dyn StylesheetOwner, document: &Document) {
+        if owner.parser_inserted() {
+            document.decrement_script_blocking_stylesheet_count();
+        }
+
+        // From <https://html.spec.whatwg.org/multipage/#link-type-stylesheet>:
+        // > A link element of this type is implicitly potentially render-blocking if the element
+        // > was created by its node document's parser.
+        if matches!(self.source, StylesheetContextSource::LinkElement) && owner.parser_inserted() {
+            document.decrement_render_blocking_element_count();
+        }
+    }
+
+    fn do_post_parse_tasks(self, successful: bool, stylesheet: Option<Arc<Stylesheet>>) {
+        let element = self.element.root();
+        let document = self.document.root();
+        let owner = element
+            .upcast::<Element>()
+            .as_stylesheet_owner()
+            .expect("Stylesheet not loaded by <style> or <link> element!");
+
+        if self.is_for_a_previous_generation(&element) {
+            self.decrement_load_and_render_blockers(owner, &document);
+        }
+
+        if let Some(stylesheet) = stylesheet {
+            match &self.source {
+                StylesheetContextSource::LinkElement => {
+                    let link = element.downcast::<HTMLLinkElement>().unwrap();
+                    if link.is_effectively_disabled() {
+                        stylesheet.set_disabled(true);
+                    }
+                    link.set_stylesheet(stylesheet);
+                },
+                StylesheetContextSource::Import { import_rule } => {
+                    // Layout knows about this stylesheet, because Stylo added it to the Stylist,
+                    // but Layout doesn't know about any new web fonts that it contains.
+                    document.load_web_fonts_from_stylesheet(&stylesheet);
+
+                    let mut guard = document.style_shared_lock().write();
+                    import_rule.write_with(&mut guard).stylesheet = ImportSheet::Sheet(stylesheet);
+                },
+            }
+        }
+
+        if let Some(ref shadow_root) = self.shadow_root {
+            shadow_root.root().invalidate_stylesheets();
+        } else {
+            document.invalidate_stylesheets();
+        }
+
+        self.decrement_load_and_render_blockers(owner, &document);
+
+        owner.set_origin_clean(self.origin_clean);
+        document.finish_load(LoadType::Stylesheet(self.url.clone()), CanGc::note());
+
+        if let Some(any_failed) = owner.load_finished(successful) {
+            let event = match any_failed {
+                true => atom!("error"),
+                false => atom!("load"),
+            };
```

Comment:
Ultra-nit: why not use if instead of match on a boolean?

```suggestion
            let event = if any_failed { atom!("error") } else { atom!("load") };
```

Reply:
I went with a match because it was one less line, but your approach is even better.

Reply:
Hrm. It seems that `rustfmt` doesn't like the single-line version, so maybe I'll stick with the `match` since it is one less line than this:

```rust
            let event = match any_failed {
                true => atom!("error"),
                false => atom!("load"),
            };
```